api/src/main/java/jakarta/enterprise/inject/build/compatible/spi/ClassConfig.java

```diff
@@ -5,23 +5,42 @@
 import java.util.Collection;
 
 /**
- * @param <T> the configured class
+ * Allows adding annotations to and removing annotations from a class.
+ * Note that the class is not physically altered, the modifications
+ * are only seen by the CDI container.
+ *
+ * @see Enhancement
  */
-public interface ClassConfig<T> extends ClassInfo<T>, AnnotationConfig {
-    // TODO remove the type parameter?
-    // TODO even if ClassInfo has equals/hashCode, ClassConfig probably shouldn't
-
-    // only constructors declared by this class, not inherited ones
-    // no [static] initializers
+public interface ClassConfig extends DeclarationConfig {
+    /**
+     * Returns the {@link ClassInfo read-only information} about this transformed class.
+     *
+     * @return the {@link ClassInfo} corresponding to this transformed class
+     */
     @Override
-    Collection<? extends MethodConfig<T>> constructors();
+    ClassInfo info();
 
-    // only methods declared by this class, not inherited ones
-    // no constructors nor [static] initializers
-    @Override
-    Collection<? extends MethodConfig<T>> methods();
+    /**
+     * Returns a collection of {@link MethodConfig} objects for each constructor of this class.
+     *
+     * @return a collection of {@link MethodConfig} objects, never {@code null}
+     */
+    // TODO specify whether inherited constructors are also included
+    Collection<MethodConfig> constructors();
 
-    // only fields declared by this class, not inherited ones
-    @Override
-    Collection<? extends FieldConfig<T>> fields();
+    /**
+     * Returns a collection of {@link MethodConfig} objects for each method of this class.
+     *
+     * @return a collection of {@link MethodConfig} objects, never {@code null}
+     */
+    // TODO specify whether inherited methods are also included
+    Collection<MethodConfig> methods();
```

Comment:
So this is something I'm a little ambivalent about. I think we already discussed something similar for annotations and kinda agreed that we'll always return all annotations, including inherited ones. We should probably do the same here.

I'm looking at what Weld does for Portable Extensions (because that seems quite under-documented on the Portable Extensions API side), and it works like this:

- for constructors, returns only constructors declared on this class
- for fields, returns all fields declared on this class and all its superclasses up to and _excluding_ `Object` (which also includes `private` fields from superclasses)
- for methods, returns all methods declared on this class and all its superclasses up to and _excluding_ `Object` (which also includes `private` methods from superclasses), plus also all `default` methods from all interfaces directly and indirectly implemented by this class

I guess that kinda sorta makes sense. WDYT?

Reply:
I think it makes sense for these methods yes. Although they will be relatively expensive methods to call since materializing all that info could be costly.

Micronaut has an extensive querying API for methods/fields that lets you filter methods prior to materializing the actual equivalent of `MethodConfig` https://github.com/micronaut-projects/micronaut-core/blob/3.0.x/inject/src/main/java/io/micronaut/inject/ast/ElementQuery.java

I am not sure if something like this makes sense here.

Reply:
I was just thinking about it and I've got two little problems with methods from interfaces. We would include `abstract` methods from [super]classes, but not from [super]interfaces. Also, we would include all `default` methods, even if they potentially clash. It still kinda makes sense, but feels arbitrary. I'll think a bit more about it.

The situation for constructors and fields is rather clear, so I'll draft something and amend the PR.

Also, we already had a bit of querying API in `AppArchive` and `AppDeployment`. We might be able to bring some of that back for sure, but I don't think this is too big of an issue. Implementations should be able to materialize the collection lazily and memoize the result.